### PR TITLE
feat(evm): add canonical erc-20 approval tx builder

### PR DESCRIPTION
**Pull request by gomesalexandre**
follow-up for vultisig/agent-backend-ts#1292

agent-backend-ts's `erc20-approve.ts` hand-rolls its own token amount parsing, code-on-chain safety checks, balance sanity bounds, and approve calldata construction locally, instead of consuming an SDK-owned canonical surface. Approval semantics are money-critical — leaving parsing/safety/calldata construction backend-local invites drift between backend/app/CLI/SDK surfaces.

Adds `packages/sdk/src/tools/evm/erc20ApprovalTx.ts`:
- `parseErc20ApprovalAmount` — handles `max`, revoke (`0`), strict base-unit strings, and human-readable amount scaling with token decimals
- `buildErc20ApprovalTx` — returns a backend-compatible unsigned envelope (`{chain, chain_id, to, value, data, spender, amount, is_unlimited}`) plus normalized metadata
- Injectable validation hooks (`hasCode`, `readDecimals`, `readBalance`) so consumers wire their own RPC plumbing
- Default validation is fund-conservative: invalid addresses fail before calldata construction, token/spender bytecode required, specific approvals bounded against balance when available, and no-code results never fail open even with transient-RPC compatibility flags enabled

Exported from `sdk.evm`, `sdk.tools`, and root `@vultisig/sdk`.

Purely additive — existing `encodeErc20Approve`/`encodeErc20Revoke`/`MAX_UINT256` behavior unchanged, no existing transaction/signing path rewired. Fund-adjacent but unsigned-only: builds an approval envelope and calldata, does not sign or broadcast.

**Scope note**: SDK-side half only. Migrating `agent-backend-ts`'s `erc20_approve` to call this (preserving its Blockaid `scan_request` wrapping) is the legitimate cross-repo follow-up, not implementable from this worktree — flagged honestly, not silently dropped. Once that migration lands, a fast-vault approve/revoke smoke test is the natural verification step.

## what I ran
```
yarn workspace @vultisig/sdk exec vitest run .../erc20ApprovalTx.test.ts .../encodeErc20Approve.test.ts .../toolNamespaceRootExports.test.ts .../publicExports.test.ts
Test Files  4 passed (4)
Tests  78 passed (78)
```
Full typecheck, eslint, prettier, `yarn build:sdk`, and `yarn changeset:check-sdk-bundle` all green. Verified real runtime behavior via direct import of the built dist (`parseErc20ApprovalAmount({amount:'1.5',decimals:6})` → `1500000` base units, correct). Also did a full CLI build + smoke (`vultisig --help`) to confirm the change doesn't break the CLI consumer. Changeset included.

Co-Authored-By: Claude <noreply@anthropic.com>

## Summary by CodeRabbit

* **New Features**
  * Added ERC-20 approval transaction creation for EVM networks.
  * Supports max, revoke, human-readable, and base-unit approval amounts.
  * Added validation for token details, addresses, balances, and optional contract checks.
  * Exposed approval builders, amount parsers, validation options, and result types through the SDK.

* **Bug Fixes**
  * Added safeguards for invalid formats, excessive precision, and unsafe approval amounts.

### CODE REVIEW

**Review by gomesalexandre — COMMENTED** [review]
preferably-blocking:
- `packages/sdk/src/tools/evm/erc20ApprovalTx.ts:181-214` silently truncates discarded fractional precision. With `amount='0.0000001'` and `decimals=6`, `fracPart.slice(0, decimals).padEnd(...)` produces `000000`, so the built approval amount becomes `0` while the metadata still comes back as a `specific` approval. That is a footgun on a money path. Please reject discarded non-zero digits and add a regression for the sub-decimal case.
- The PR body says `closes vultisig/agent-backend-ts#1292`, but the backend still owns a separate `erc20_approve` implementation today. Please drop the `closes` claim or land the consumer migration in the same rollout.

**Comment by gomesalexandre** [comment]
pushed 3765bd23340e2cffaa29386662aa1c08f266b6f9 to address the current blocker.

what changed
- `parseHumanTokenAmount` now rejects discarded non-zero fractional digits beyond token decimals instead of silently truncating them to zero or another lower amount
- added regression coverage for both `100.1234567` and `0.0000001` at 6 decimals
- also clarified the PR body from `closes vultisig/agent-backend-ts#1292` to `follow-up for vultisig/agent-backend-ts#1292`

receipt
- `export PATH="/Users/gomes/.nvm/versions/node/v22.22.0/bin:$PATH" && ./node_modules/.bin/vitest run --config packages/sdk/tests/unit/vitest.config.ts packages/sdk/tests/unit/tools/evm/erc20ApprovalTx.test.ts` -> 1 file passed, 13 tests passed
- `YARN_ENABLE_IMMUTABLE_INSTALLS=false yarn build:shared` is still blocked the same way as pristine main by the pre-existing cardano `auxiliaryData` typing drift, so i am not claiming a new full-build receipt from this checkout

fresh current-head review wanted once ci reruns.

File: packages/sdk/src/tools/evm/erc20ApprovalTx.ts [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    wholePart = trimmed.slice(0, dotIdx)
+    fracPart = trimmed.slice(dotIdx + 1)
+    if (fracPart.includes('.')) {
+      throw new Error(`invalid amount: multiple decimal points in ${amount}`)
+    }
+  }
+
+  if (wholePart === '') {
+    wholePart = '0'
+  }
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/sdk/src/tools/evm/erc20ApprovalTx.ts]: fixed in 264fef1d8.

- merged latest origin/main into this branch and resolved the tools index export conflict by keeping both the new erc20 approval exports and main's agent-router checkout exports
- rejected bare `"."` approval amounts so they no longer collapse into a zero-value revoke path
- added a regression in `packages/sdk/tests/unit/tools/evm/erc20ApprovalTx.test.ts`

receipts:
- `vitest run --config tests/unit/vitest.config.ts tests/unit/tools/evm/erc20ApprovalTx.test.ts` ✅
- `vitest run --config tests/unit/vitest.config.ts tests/unit/utils/publicExports.test.ts` still hit an environment-tainted borrowed-install export error around `@scure/bip39` after the mergefix, so current-head ci still needs to be the authoritative public-export receipt

**Comment by gomesalexandre** [comment]
pushed 264fef1d8 to `fix/tl-sdk-1292`.

what changed
- merged fresh `origin/main` into the branch and cleared the `packages/sdk/src/tools/index.ts` conflict by preserving both sides of the export surface
- fixed the new approval parser edge case so a bare `"."` is rejected instead of being normalized into a revoke amount
- added a regression test for that input

receipts
- `vitest run --config tests/unit/vitest.config.ts tests/unit/tools/evm/erc20ApprovalTx.test.ts` ✅
- `vitest run --config tests/unit/vitest.config.ts tests/unit/utils/publicExports.test.ts` hit a borrowed-install/export-environment issue at `@scure/bip39`, so i'm leaving current-head ci as the authoritative public-export receipt here
- `yarn build:sdk` also stops on an unrelated cardano typing error (`packages/core/mpc/keysign/signingInputs/resolvers/cardano.ts:72 auxiliaryData`) after the merge

head changed, so any old approval is stale. needs a fresh non-gomes review on the new head before merge.

**Review by rcoderdev — APPROVED** [review]

**Comment by rcoderdev** [comment]
Closing this because The linked backend still has its own ERC-20 approval implementation, and no current SDK or Windows consumer has been demonstrated for this extra public builder. Main has a React Native approval builder. This conflicting branch needs another integration and QA cycle; please return with a concrete SDK consumer requirement if the general helper is needed.

This PR has been open for more than two weeks, and we do not plan to carry this version forward.

If the change is still needed, please open a fresh PR against current main with the updated context.
